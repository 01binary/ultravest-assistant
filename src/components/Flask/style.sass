.flask {
    border: 1px solid gray;
}