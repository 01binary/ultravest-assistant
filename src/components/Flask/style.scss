@import '../App/style/colors';
@import '../App/style/layout';
@import '../App/style/timeline';

// Timeline color
.flask:after {
	border-color: $color-primary;
}

// Timeline icon
.flask {
	background: url('../../assets/flask.svg') 0 50% no-repeat;
	background-size: $timeline-icon-size $timeline-icon-size;
}

// Height label smaller and with margin
.dimensions section:last-of-type label {
	margin-left: $spacing-half;
	flex: 0 0 3.5em;
}

// Units
.dimensions span {
	margin: 1em 0 0 calc(-1em - 8px);
	order: -2;
	opacity: .5;
	transition: opacity .3s ease-in-out;
}

// Hide units on hover or focus to make way for up/down spinner
.dimensions input:focus ~ span,
.dimensions section:hover span {
	opacity: 0;
}

// Buttons
.action {
	flex: 0 0 $spacing-double;
	background-color: $color-border-dark;
	border-radius: $spacing-quarter;
	border: none;
	width: $spacing-double;
	height: $spacing-double;
	margin-left: $spacing-third;
	cursor: pointer;
	transition:
		color .2s ease-in-out,
		background-color .2s ease-in-out;
	appearance: none;
	-moz-appearance: none;
	-webkit-appearance: none;
	text-indent: -9999px;
	clip: rect(1px 1px 1px 1px); /* IE6, IE7 */
	clip: rect(1px, 1px, 1px, 1px);

	&:hover,
	&:focus {
		color: $color-text-darker;
		background-color: $color-border-light;
	}
}

.add {
	background-image: url('../../assets/add.svg');
}

.remove {
	background-image: url('../../assets/remove.svg');
}
