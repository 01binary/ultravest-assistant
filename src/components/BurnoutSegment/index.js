import { h } from 'preact';
import style from './style';

/**
 * Burnout diagram secment.
 * @param {*} props 
 */
const BurnoutSegment = ({  }) => (
	<article>

	</article>
);

export default BurnoutSegment;