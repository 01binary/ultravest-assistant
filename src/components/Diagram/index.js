import { h } from 'preact';
import Segment from '../Segment';
import getSegmentOffset from '../../selectors/getSegmentOffset';
import style from './style';

/**
 * Burnout diagram
 * @param {object[]} segments - The program segments.
 * @param {number} maxIndex - The index of the segment with the highest temperature.
 * @returns {JSX.Element} - A stateless component.
 */
const Diagram = ({
	segments,
	maxIndex }) => (
	<section class={style.diagram}>
		{ segments.map((segment, index) => (
			<Segment
				{...segment}
				prev={segments[index - 1]}
				offset={getSegmentOffset(index, maxIndex, segments)}
			/>
		))}
	</section>
);

export default Diagram;
