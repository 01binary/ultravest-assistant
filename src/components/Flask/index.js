import { h } from 'preact';
import { CUSTOM_PRESET } from '../../enhancers/withFlask';
import style from './style';

/**
 * Flask Parameters.
 * @param {number} props.flask - The flask props.
 * @param {func} props.handleFlaskPresetChange - The function called to set flask preset.
 * @param {func} props.handleFlaskDiameterChange - The function called to set flask diameter.
 * @param {func} props.handleFlaskHeightChange - The function called to set flask height.
 * @returns {JSX.Element} - A React stateless component.
 */
const Flask = (
	{
		flask,
		handleFlaskPresetChange,
		handleFlaskDiameterChange,
		handleFlaskHeightChange
	}) => (

	<article class={style.flask}>
		<h2>flask</h2>

		<label for="flask-preset">
			preset
		</label>

		<select
			name="flask-preset"
			value={flask.preset}
			onChange={handleFlaskPresetChange}
		>
			{ Object.keys(flask.presets).map(preset => (
				<option selected={preset === flask.preset}>
					{preset}
				</option>
			))}

			<option selected={flask.preset === CUSTOM_PRESET}>
				Custom
			</option>
		</select>

		<fieldset>
			<label for="flask-diameter">diameter</label>
			<input
				name="flask-diameter"
				type="number"
				step="0.1"
				value={flask.diameter}
				onChange={handleFlaskDiameterChange}
			/>

			<label for="flask-height">height</label>
			<input
				name="flask-height"
				type="number"
				step="0.1"
				value={flask.height}
				onChange={handleFlaskHeightChange}
			/>
		</fieldset>
	</article>
);

export default Flask;
