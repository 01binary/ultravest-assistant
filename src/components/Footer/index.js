import { h } from 'preact';
import './style';

const Footer = () => (
	<footer>
		<a
			href="https://github.com/01binary/ultravest-assistant"
			target="_blank"
			rel="noopener noreferrer"
		>
			fork on <strong>GitHub</strong>
		</a>
	</footer>
);

export default Footer;
