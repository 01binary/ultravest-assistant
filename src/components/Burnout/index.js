import { h } from 'preact';
import presets from '../../config/burnoutPresets';
import getBurnoutPreset from '../../selectors/getBurnoutPreset';
import getBurnoutTime from '../../selectors/getBurnoutTime';
import Diagram from '../Diagram';
import Steps from '../Steps';
import style from './style';

/**
 * Display burnout preset calculated from flask size.
 * @param {object} flask - The flask props.
 * @param {bool} showSegments - Whether to show raw or formatted burnout steps.
 * @param {func} handleChangeView - Switches between raw and formatted steps.
 * @returns {JSX.Element} - A stateless component.
 */
const Burnout = ({
	flask,
	showSegments,
	handleChangeView
}) => {
	const preset = getBurnoutPreset(presets, flask);

	return (
		<article class={style.burnout}>
			<h2>burnout</h2>

			<output>
				<dl>
					<dt>preset</dt>
					<dd>
						{preset.diameter} X {preset.height} in
						{ getIsLastPreset(preset) &&
							<span class={style.longest}>longest</span>
						}
					</dd>

					<dt>time</dt>
					<dd>{getBurnoutTime(preset)} hours</dd>
				</dl>
			</output>

			<Diagram segments={preset.segments} />

			<Steps
				segments={preset.segments}
				showSegments={showSegments}
				handleChangeView={handleChangeView}
			/>
		</article>
	);
};

const getIsLastPreset = preset => (
	preset === presets[presets.length - 1] || null
);

export default Burnout;
