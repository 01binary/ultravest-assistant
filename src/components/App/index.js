
import { h } from 'preact';
import Header from '../Header';
import Flask from '../Flask';
import Investment from '../Investment';
import Burnout from '../Burnout';
import Footer from '../Footer';
import getFlask from '../../selectors/getFlask';
import style from './style';

/**
 * Ultravest Assistant App
 * @param {object} props - The composed application state.
 * @returns {function} - A stateless component.
 */
const App = props => (
	<form id="app" class={style.app}>
		<Header />

		<main>
			<Flask {...getFlask(props)} />
			<Investment {...props} />
			<Burnout {...props} />
		</main>

		<Footer />
	</form>
);

export default App;
