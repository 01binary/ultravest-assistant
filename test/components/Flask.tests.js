import { h } from 'preact';
import { deep } from 'preact-render-spy';
import { Flask } from '../../src/components/Flask';

describe('Flask', () => {

	let wrapper, props;

	beforeAll(() => {
		props = {
			query: {},
			preset: expectedName,
			presets: {
				[expectedName]: expected,
				'another preset': {
					diameter: 4,
					height: 6
				}
			},
			diameter: expected.diameter,
			height: expected.height,
			handleFlaskPresetChange: jest.fn(),
			handleFlaskDiameterChange: jest.fn(),
			handleFlaskHeightChange: jest.fn(),
			handleAddFlaskPreset: jest.fn(),
			handleRemoveFlaskPreset: jest.fn()
		};

		wrapper = deep(
			<Flask {...props} />
		);
	});

	afterAll(() => {
		wrapper = null;
	});

	it('should render', () => {
		expect(wrapper).toMatchSnapshot();
	});

	it('should bind preset value', () => {
		const first = wrapper.find('option')[0];
		expect(first.children[0]).toBe(expectedName);
		expect(first.attributes.selected).toBe(true);
	});

	it('should bind diameter', () => {
		expect(wrapper.find('#flask-diameter')[0].attributes.value)
			.toBe(expected.diameter);
	});

	it('should bind height', () => {
		expect(wrapper.find('#flask-height')[0].attributes.value)
			.toBe(expected.height);
	});

	it('should bind change preset handler', () => {
		wrapper.find('#flask-preset')[0].attributes.onChange();
		expect(props.handleFlaskPresetChange.mock.calls.length)
			.toBe(1);
	});

	it('should bind change diameter handler', () => {
		wrapper.find('#flask-diameter')[0].attributes.onChange();
		expect(props.handleFlaskDiameterChange.mock.calls.length)
			.toBe(1);
	});

	it('should bind change height handler', () => {
		wrapper.find('#flask-height')[0].attributes.onChange();
		expect(props.handleFlaskHeightChange.mock.calls.length)
			.toBe(1);
	});

	it('should bind add preset handler', () => {
		wrapper.find('#action-add')[0].attributes.onClick();
		expect(props.handleAddFlaskPreset.mock.calls.length)
			.toBe(1);
	});

	it('should bind remove preset handler', () => {
		wrapper.find('#action-remove')[0].attributes.onClick();
		expect(props.handleRemoveFlaskPreset.mock.calls.length)
			.toBe(1);
	});

	const expectedName = 'test preset';
	const expected = {
		diameter: 2,
		height: 3
	};
});
