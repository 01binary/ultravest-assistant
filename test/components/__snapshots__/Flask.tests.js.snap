// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Flask should render 1`] = `
preact-render-spy (1 nodes)
-------
<article class="flask">
  <h2>flask</h2>
  <label for="flask-preset">preset</label>
  <select
    name="flask-preset"
    value="3.5 X 10"
    onChange={[Function anonymous]}
  >
    <option selected={false}>3.5 X 5</option>
    <option selected={true}>3.5 X 10</option>
    <option selected={false}>4 x 10</option>
    <option selected={false}>4 x 14</option>
    <option selected={false}>Custom</option>
  </select>
  <fieldset>
    <label for="flask-diameter">diameter</label>
    <input
      name="flask-diameter"
      type="number"
      step="0.1"
      value={3.5}
      onChange={[Function anonymous]}
     />
    </input>
    <label for="flask-height">height</label>
    <input
      name="flask-height"
      type="number"
      step="0.1"
      value={10}
      onChange={[Function anonymous]}
     />
    </input>
  </fieldset>
</article>

`;
